### PR TITLE
DType regression

### PR DESCRIPTION
**Pull request by Godricly**
* DType regression
* Python test for DType pooling and regression

### CODE REVIEW

File: src/operator/regression_output-inl.h [hunk 1]
```diff
@@ -69,11 +69,11 @@ class RegressionOutputOp : public Operator {
...
-    Tensor<xpu, 2> out = out_data[reg_enum::kOut].FlatTo2D<xpu, real_t>(s);
-    Tensor<xpu, 2> grad = in_grad[reg_enum::kData].FlatTo2D<xpu, real_t>(s);
-    Tensor<xpu, 2> label = in_data[reg_enum::kLabel]
-      .get_with_shape<xpu, 2, real_t>(out.shape_, s);
-    Assign(grad, req[reg_enum::kData], param_.grad_scale/num_output*
+    Tensor<xpu, 2, DType> out = out_data[reg_enum::kOut].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> grad = in_grad[reg_enum::kData].FlatTo2D<xpu, DType>(s);
+    Tensor<xpu, 2, DType> label = in_data[reg_enum::kLabel]
+      .get_with_shape<xpu, 2, DType>(out.shape_, s);
+    Assign(grad, req[reg_enum::kData], scalar<DType>(param_.grad_scale/num_output)*
```
Comment (winstywang) [inline src/operator/regression_output-inl.h]: Is this correct? What if Dtype is int8?

File: src/operator/regression_output-inl.h [hunk 2]
```diff
@@ -129,6 +129,27 @@ class RegressionOutputProp : public OperatorProperty {
...
+    bool InferType(std::vector<int> *in_type,
+                 std::vector<int> *out_type,
+                 std::vector<int> *aux_type) const override {
+    CHECK_EQ(in_type->size(), 2) << "Input:[data, label]";
+    int dtype = (*in_type)[0];
+
+    auto nin = in_type->size();
+    in_type->clear();
+    in_type->push_back(dtype);
+    for (index_t i = 1; i < nin; ++i) in_type->push_back(dtype);
```
Comment (winstywang) [inline src/operator/regression_output-inl.h]: Please follow the coding convention, use {} to wrap it, and put it in a separate line.

File: src/operator/regression_output.cc [hunk 3]
```diff
@@ -11,24 +11,38 @@ namespace op {
...
-    case reg_enum::kLogistic:
-      return new RegressionOutputOp<cpu, mshadow_op::sigmoid, mshadow::op::minus>(param);
-    case reg_enum::kMAE:
-      return new RegressionOutputOp<cpu, mshadow::op::identity, mshadow_op::minus_sign>(param);
-    default:
-      LOG(FATAL) << "unknown activation type " << type;
-  }
-  return nullptr;
+                                        RegressionOutputParam param, int dtype) {
+  Operator *op = NULL;
```
Comment (winstywang) [inline src/operator/regression_output.cc]: use nullptr?

File: src/operator/regression_output-inl.h [hunk 4]
```diff
@@ -165,7 +186,13 @@ class RegressionOutputProp : public OperatorProperty {
     return {{in_data[reg_enum::kData], out_data[reg_enum::kOut]}};
   }
 
-  Operator* CreateOperator(Context ctx) const override;
+  Operator* CreateOperator(Context ctx) const override {
+    LOG(FATAL) << "Not Implemented.";
+    return NULL;
```
Comment (winstywang) [inline src/operator/regression_output-inl.h]: nullptr

**Comment by winstywang** [comment]
Some inline comments are added. After fixing them, it is good to merge in

File: src/operator/regression_output-inl.h (hunk 1 above)
Comment (Godricly) [inline src/operator/regression_output-inl.h]: Thank you for your review. Which line are you worry about? I think int8 is currently not supported as we are using `MSHADOW_REAL_TYPE_SWITCH` to create operators.

Comment (winstywang) [inline src/operator/regression_output-inl.h]: OK, merge as is for now. We have to change a lot for the coming unit8...

Comment (Godricly) [inline src/operator/regression_output-inl.h]: yep... I'll propose a pr to fix nullptr issue in other operators later. Is there any paper list to share about uint8 network?

There are some other issues to be fixed to make DType network really works, like kvstore, param init. For the lstm case, extra states data_type is need.  We need to discuss more about this DType support and a more detail todo list.

**Comment by winstywang** [comment]
Please update to latest master, so that I can merge

**Comment by Godricly** [comment]
Cpp test failed after the update.

**Comment by winstywang** [comment]
@Godricly We have to revert this PR until you provide correct test scripts and corresponding results.

**Comment by Godricly** [comment]
Sry for that typo. I don't have GPU access now,  so a fix will take some time. The operator should be ok. Does that script get called in travis?

**Pull request merged**